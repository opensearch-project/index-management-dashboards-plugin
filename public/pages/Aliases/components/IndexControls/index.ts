/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import IndexControls from "./IndexControls";

export * from "./IndexControls";
export default IndexControls;
