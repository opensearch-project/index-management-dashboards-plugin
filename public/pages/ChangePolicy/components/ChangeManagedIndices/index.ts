/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ChangeManagedIndices from "./ChangeManagedIndices";

export default ChangeManagedIndices;
