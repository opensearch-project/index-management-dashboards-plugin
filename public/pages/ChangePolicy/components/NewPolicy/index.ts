/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import NewPolicy from "./NewPolicy";

export default NewPolicy;
