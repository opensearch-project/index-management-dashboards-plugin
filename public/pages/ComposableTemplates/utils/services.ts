/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { CommonService } from "../../../services";
import { TemplateItemRemote } from "../../../../models/interfaces";

export const getAllUsedComponents = async ({ commonService }: { commonService: CommonService }) => {
  const allTemplatesResponse = await commonService.apiCaller<{
    index_templates?: Array<{
      name: string;
      index_template: TemplateItemRemote;
    }>;
  }>({
    endpoint: "transport.request",
    data: {
      method: "GET",
      path: "/_index_template/*",
    },
  });

  if (allTemplatesResponse && allTemplatesResponse.ok) {
    return allTemplatesResponse.response.index_templates || [];
  }

  return [];
};
