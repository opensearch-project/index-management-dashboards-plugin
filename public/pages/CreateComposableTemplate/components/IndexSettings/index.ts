import IndexSettings from "./IndexSettings";

export default IndexSettings;
