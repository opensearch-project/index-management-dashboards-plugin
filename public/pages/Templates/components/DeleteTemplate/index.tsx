/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import React, { useState } from "react";
import { EuiButtonIcon, EuiToolTip } from "@elastic/eui";
import DeleteIndexModal from "../../containers/DeleteTemplatesModal";
import { ITemplate } from "../../interface";

export interface DeleteTemplateProps {
  selectedItems: ITemplate[];
  onDelete: () => void;
}

export default function DeleteTemplate(props: DeleteTemplateProps) {
  const { selectedItems, onDelete } = props;
  const [deleteIndexModalVisible, setDeleteIndexModalVisible] = useState(false);

  const onDeleteIndexModalClose = () => {
    setDeleteIndexModalVisible(false);
  };

  return (
    <>
      <EuiToolTip content="Delete template">
        <EuiButtonIcon
          aria-label="Delete template"
          color="danger"
          iconType="trash"
          onClick={() => setDeleteIndexModalVisible(true)}
          className="icon-hover-danger"
        />
      </EuiToolTip>
      <DeleteIndexModal
        selectedItems={selectedItems.map((item) => item.name)}
        visible={deleteIndexModalVisible}
        onClose={onDeleteIndexModalClose}
        onDelete={() => {
          onDeleteIndexModalClose();
          onDelete();
        }}
      />
    </>
  );
}
