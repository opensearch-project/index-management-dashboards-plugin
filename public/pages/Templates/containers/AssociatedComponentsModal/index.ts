/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import AssociatedComponentsModal from "./AssociatedComponentsModal";

export default AssociatedComponentsModal;
