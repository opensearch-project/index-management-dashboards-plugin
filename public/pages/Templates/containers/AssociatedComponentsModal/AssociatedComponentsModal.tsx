/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { Dispatch, SetStateAction, useContext, useState } from "react";
import { CoreStart } from "opensearch-dashboards/public";
import { ServicesContext } from "../../../../services";
import { CoreServicesContext } from "../../../../components/core_services";
import {
  EuiSmallButtonIcon,
  EuiFlyout,
  EuiFlyoutBody,
  EuiFlyoutHeader,
  EuiInMemoryTable,
  EuiLink,
  EuiTitle,
  EuiToolTip,
  EuiText,
} from "@elastic/eui";
import { ROUTES } from "../../../../utils/constants";
import { ReactChild } from "react";
import { Modal } from "../../../../components/Modal";
import { BrowserServices } from "../../../../models/interfaces";
import { TemplateItemRemote } from "../../../../../models/interfaces";
import { ITemplate } from "../../interface";
import { getTemplate } from "../../../ComposableTemplates/utils/hooks";

export interface AssociatedComponentsModalProps {
  template: ITemplate;
  onUnlink?: (unlinkTemplate: string) => void;
  renderProps: (params: { setVisible: Dispatch<SetStateAction<boolean>> }) => ReactChild;
}

export default function AssociatedComponentsModal(props: AssociatedComponentsModalProps) {
  const { onUnlink, renderProps, template } = props;
  const [visible, setVisible] = useState(false);
  const services = useContext(ServicesContext) as BrowserServices;
  const coreServices = useContext(CoreServicesContext) as CoreStart;

  return (
    <>
      {renderProps({ setVisible })}
      {visible ? (
        <EuiFlyout onClose={() => setVisible(false)}>
          <EuiFlyoutHeader>
            <EuiText size="s">
              <h2>Associated component templates</h2>
            </EuiText>
          </EuiFlyoutHeader>
          <EuiFlyoutBody>
            <EuiInMemoryTable
              pagination
              items={(template.templateDetail?.composed_of || []).map((item) => ({
                name: item,
              }))}
              columns={[
                {
                  name: "Component template",
                  field: "name",
                  sortable: true,
                  render: (value: string, record) => (
                    <EuiLink external={false} target="_blank" href={`#${ROUTES.CREATE_COMPOSABLE_TEMPLATE}/${value}`}>
                      {value}
                    </EuiLink>
                  ),
                },
                {
                  name: "Actions",
                  field: "actions",
                  align: "right",
                  render: (value: string, record) => {
                    return (
                      <EuiToolTip content="Unlink">
                        <EuiSmallButtonIcon
                          aria-label={`Unlink ${record.name}?`}
                          iconType="unlink"
                          onClick={() => {
                            Modal.show({
                              type: "confirm",
                              title: `Unlink from ${template.name}?`,
                              content: (
                                <p style={{ lineHeight: 1.5 }}>
                                  The component {record.name} will be removed from the template {template.name}. This will affect any new
                                  indexes created with this template.
                                </p>
                              ),
                              footer: ["cancel", "confirm"],
                              locale: {
                                confirm: "Unlink",
                              },
                              confirmButtonProps: {
                                color: "danger",
                              },
                              async onOk() {
                                const currentTemplate = await getTemplate({
                                  templateName: template.name,
                                  commonService: services.commonService,
                                  coreService: coreServices,
                                });
                                const updateResult = await services.commonService.apiCaller({
                                  endpoint: "transport.request",
                                  data: {
                                    method: "POST",
                                    path: `/_index_template/${template.name}`,
                                    body: {
                                      ...currentTemplate,
                                      composed_of: currentTemplate?.composed_of?.filter((item) => item !== record.name) || [],
                                    } as TemplateItemRemote,
                                  },
                                });
                                if (updateResult.ok) {
                                  onUnlink?.(record.name);
                                  coreServices.notifications.toasts.addSuccess(
                                    `${record.name} has been successfully unlinked from ${template.name}.`
                                  );
                                } else {
                                  coreServices.notifications.toasts.addDanger(updateResult.error);
                                  return Promise.reject(updateResult.error);
                                }
                              },
                            });
                          }}
                        />
                      </EuiToolTip>
                    );
                  },
                },
              ]}
            />
          </EuiFlyoutBody>
        </EuiFlyout>
      ) : null}
    </>
  );
}
