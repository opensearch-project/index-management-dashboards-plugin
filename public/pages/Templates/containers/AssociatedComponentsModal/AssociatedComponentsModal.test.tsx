/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import AssociatedComponentsModal, { AssociatedComponentsModalProps } from "./AssociatedComponentsModal";
import { ServicesContext } from "../../../../services";
import { CoreServicesContext } from "../../../../components/core_services";
import { browserServicesMock, coreServicesMock } from "../../../../../test/mocks";

function renderWithServices(props: AssociatedComponentsModalProps) {
  return {
    ...render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <ServicesContext.Provider value={browserServicesMock}>
          <AssociatedComponentsModal {...props} />
        </ServicesContext.Provider>
      </CoreServicesContext.Provider>
    ),
  };
}

describe("<AssociatedComponentsModal /> spec", () => {
  it("renders with empty", async () => {
    renderWithServices({
      template: {
        name: "test_template",
        index_patterns: "['*']",
        order: 0,
        composed_of: "",
      },
      renderProps() {
        return <></>;
      },
    });
    expect(document.body.children).toMatchSnapshot();
  });

  it("renders the component", async () => {
    let time = 0;
    browserServicesMock.commonService.apiCaller = jest.fn(
      async (payload): Promise<any> => {
        if (payload.data?.path?.startsWith("/_component_template/")) {
          return {
            ok: true,
            response: {},
          };
        } else if (payload.data?.path?.startsWith("/_index_template")) {
          if (payload.data?.method === "POST") {
            time++;
            return {
              ok: time !== 1,
              error: "error",
            };
          }
          return {
            ok: true,
            response: {
              index_templates: [
                {
                  name: "test_template",
                  index_template: {
                    composed_of: ["test_component_template"],
                  },
                },
              ],
            },
          };
        }
        return { ok: true, response: {} };
      }
    );
    const unlinkHandlerMock = jest.fn();
    const { findByText, findByTestId, getByTestId, queryByText } = renderWithServices({
      template: {
        name: "test_template",
        index_patterns: "['*']",
        order: 0,
        composed_of: "",
        templateDetail: {
          composed_of: ["test_component_template"],
          template: {},
          name: "test_template",
          version: 0,
          priority: 0,
          index_patterns: ["*"],
        },
      },
      renderProps: ({ setVisible }) => (
        <button data-test-subj="TestBtn" onClick={() => setVisible(true)}>
          123
        </button>
      ),
      onUnlink: unlinkHandlerMock,
    });
    await findByTestId("TestBtn");
    await userEvent.click(getByTestId("TestBtn"));
    await findByText("Associated component templates");
    expect(document.body.children).toMatchSnapshot();
    await userEvent.click(document.querySelector(`[aria-label="Unlink test_component_template?"]`) as Element);
    await findByText("Unlink from test_template?");
    await userEvent.click(getByTestId("Unlink from test_template?-confirm"));
    await waitFor(() => {
      expect(browserServicesMock.commonService.apiCaller).toBeCalledTimes(2);
      expect(coreServicesMock.notifications.toasts.addDanger).toBeCalledWith("error");
      expect(unlinkHandlerMock).toBeCalledTimes(0);
    });
    await userEvent.click(getByTestId("Unlink from test_template?-confirm"));
    await waitFor(() => {
      expect(browserServicesMock.commonService.apiCaller).toBeCalledTimes(4);
      expect(coreServicesMock.notifications.toasts.addSuccess).toBeCalledWith(
        "test_component_template has been successfully unlinked from test_template."
      );
      expect(unlinkHandlerMock).toBeCalledWith("test_component_template");
    });
    await userEvent.click(getByTestId("euiFlyoutCloseButton"));
    await waitFor(() => {
      expect(queryByText("Associated component templates")).toBeNull();
    });
  });
});
