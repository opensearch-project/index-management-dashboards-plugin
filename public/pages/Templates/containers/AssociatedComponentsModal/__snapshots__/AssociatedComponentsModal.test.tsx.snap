// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AssociatedComponentsModal /> spec renders the component 1`] = `
HTMLCollection [
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  >
    <button
      data-test-subj="TestBtn"
    >
      123
    </button>
  </div>,
  <div
    class="euiOverlayMask euiOverlayMask--belowHeader"
  >
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        class="euiFlyout euiFlyout--medium euiFlyout--paddingLarge"
        role="dialog"
        tabindex="-1"
      >
        <button
          aria-label="Close this dialog"
          class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall euiFlyout__closeButton euiFlyout__closeButton--inside"
          data-test-subj="euiFlyoutCloseButton"
          type="button"
        >
          EuiIconMock
        </button>
        <div
          class="euiFlyoutHeader"
        >
          <h2
            class="euiTitle euiTitle--medium"
          >
            Associated component templates
          </h2>
        </div>
        <div
          class="euiFlyoutBody"
        >
          <div
            class="euiFlyoutBody__overflow"
            tabindex="0"
          >
            <div
              class="euiFlyoutBody__overflowContent"
            >
              <div
                class="euiBasicTable"
              >
                <div>
                  <div
                    class="euiTableHeaderMobile"
                  >
                    <div
                      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsBaseline euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
                    >
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      />
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      />
                    </div>
                  </div>
                  <table
                    class="euiTable euiTable--responsive"
                    id="some_html_id"
                    tabindex="-1"
                  >
                    <caption
                      class="euiScreenReaderOnly euiTableCaption"
                    />
                    <thead>
                      <tr>
                        <th
                          class="euiTableHeaderCell"
                          data-test-subj="tableHeaderCell_name_0"
                          role="columnheader"
                          scope="col"
                        >
                          <span
                            class="euiTableCellContent"
                          >
                            <span
                              class="euiTableCellContent__text"
                              title="component template"
                            >
                              component template
                            </span>
                          </span>
                        </th>
                        <th
                          class="euiTableHeaderCell"
                          data-test-subj="tableHeaderCell_actions_1"
                          role="columnheader"
                          scope="col"
                        >
                          <span
                            class="euiTableCellContent euiTableCellContent--alignRight"
                          >
                            <span
                              class="euiTableCellContent__text"
                              title="Actions"
                            >
                              Actions
                            </span>
                          </span>
                        </th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr
                        class="euiTableRow"
                      >
                        <td
                          class="euiTableRowCell"
                        >
                          <div
                            class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                          >
                            component template
                          </div>
                          <div
                            class="euiTableCellContent euiTableCellContent--overflowingContent"
                          >
                            <a
                              class="euiLink euiLink--primary"
                              href="#/create-composable-template/test_component_template"
                              rel="noopener noreferrer"
                              target="_blank"
                            >
                              test_component_template
                              <span
                                class="euiScreenReaderOnly"
                              >
                                (opens in a new tab or window)
                              </span>
                            </a>
                          </div>
                        </td>
                        <td
                          class="euiTableRowCell"
                        >
                          <div
                            class="euiTableRowCell__mobileHeader euiTableRowCell--hideForDesktop"
                          >
                            Actions
                          </div>
                          <div
                            class="euiTableCellContent euiTableCellContent--alignRight euiTableCellContent--overflowingContent"
                          >
                            <span
                              class="euiToolTipAnchor"
                            >
                              <button
                                aria-label="Unlink test_component_template?"
                                class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--xSmall"
                                type="button"
                              >
                                EuiIconMock
                              </button>
                            </span>
                          </div>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
                <div>
                  <div
                    class="euiSpacer euiSpacer--m"
                  />
                  <div
                    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
                  >
                    <div
                      class="euiFlexItem euiFlexItem--flexGrowZero"
                    >
                      <div
                        class="euiPopover euiPopover--anchorUpRight"
                      >
                        <div
                          class="euiPopover__anchor"
                        >
                          <button
                            class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall"
                            data-test-subj="tablePaginationPopoverButton"
                            type="button"
                          >
                            <span
                              class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
                            >
                              EuiIconMock
                              <span
                                class="euiButtonEmpty__text"
                              >
                                Rows per page
                                : 
                                10
                              </span>
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                    <div
                      class="euiFlexItem euiFlexItem--flexGrowZero"
                    >
                      <nav
                        class="euiPagination"
                      >
                        <button
                          aria-label="Previous page"
                          class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall"
                          data-test-subj="pagination-button-previous"
                          disabled=""
                          type="button"
                        >
                          EuiIconMock
                        </button>
                        <ul
                          class="euiPagination__list"
                        >
                          <li
                            class="euiPagination__item"
                          >
                            <button
                              aria-controls="some_html_id"
                              aria-current="true"
                              aria-label="Page 1 of 1"
                              class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--small euiButtonEmpty-isDisabled euiPaginationButton euiPaginationButton-isActive euiPaginationButton--hideOnMobile"
                              data-test-subj="pagination-button-0"
                              disabled=""
                              type="button"
                            >
                              <span
                                class="euiButtonContent euiButtonEmpty__content"
                              >
                                <span
                                  class="euiButtonEmpty__text"
                                >
                                  1
                                </span>
                              </span>
                            </button>
                          </li>
                        </ul>
                        <button
                          aria-label="Next page"
                          class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall"
                          data-test-subj="pagination-button-next"
                          disabled=""
                          type="button"
                        >
                          EuiIconMock
                        </button>
                      </nav>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>,
]
`;

exports[`<AssociatedComponentsModal /> spec renders with empty 1`] = `
HTMLCollection [
  <div />,
]
`;
